lib/Dialect/RTL/RTLTypes.cpp

```diff
@@ -23,16 +25,38 @@ using namespace circt::rtl;
 #define GET_TYPEDEF_CLASSES
 #include "circt/Dialect/RTL/RTLTypes.cpp.inc"
 
+//===----------------------------------------------------------------------===//
+// Type Helpers
+//===----------------------------------------------------------------------===/
+
+void circt::rtl::getStructInnerTypes(StructType stype,
+                                     SmallVectorImpl<Type> &types) {
+  for (const auto &field : stype.getElements())
+    types.push_back(field.type);
+}
+
+Type circt::rtl::getStructFieldType(StructType stype, StringRef fieldName) {
+  for (const auto &field : stype.getElements())
+    if (field.name == fieldName)
+      return field.type;
+  return Type();
+}
+
 /// Return true if the specified type can be used as an RTL value type, that is
 /// the set of types that can be composed together to represent synthesized,
 /// hardware but not marker types like InOutType.
-bool isRTLValueType(Type type) {
+bool circt::rtl::isRTLValueType(Type type) {
   if (auto intType = type.dyn_cast<IntegerType>())
     return intType.isSignless();
 
   if (auto array = type.dyn_cast<ArrayType>())
     return isRTLValueType(array.getElementType());
 
+  if (auto t = type.dyn_cast<StructType>()) {
+    return std::all_of(t.getElements().begin(), t.getElements().end(),
+                       [](const auto &f) { return isRTLValueType(f.type); });
+  }
```

Comment:
It might be worthwhile to cache this bit in StructType to avoid compile time blowup with deep nested structs.  Alternatively, you could just out-right ban structs with inout types in them, which would make it always true for them.  That is probably the right formulation of this check for both arrays and structs, and would help solve the 'rtl types should be extensible, except not allow inout'.

It is fine to split this out to a separate patch as it is logically different than what you're working on here.  If you do, please file an issue so we don't lose it

Reply:
#391 
I think we want both the no-inout-type and no-non-rtl-type bits.

Reply:
The RTL dialect intentionally has an open type system, so it should be just no-inout-type and allow the rest.